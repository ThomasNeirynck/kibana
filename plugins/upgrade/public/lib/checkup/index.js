import {
  getFromApi,
} from '../request';


export async function getDeprecations() {
  return await getFromApi(`/api/migration/deprecations`);
}
