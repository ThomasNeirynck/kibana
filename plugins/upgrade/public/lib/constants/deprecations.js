export const DEPRECATION_ISSUE_LEVELS = [
  'critical',
  'warning',
  'info',
  'none',
];
