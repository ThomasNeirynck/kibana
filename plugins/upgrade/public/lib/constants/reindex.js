export const REINDEX_STEP_SEQUENCE = [
  'CREATE_INDEX',
  'SET_READONLY',
  'REINDEX',
  'REFRESH_INDEX',
  'VERIFY_DOCS',
  'REPLACE_INDEX',
];

export const REINDEX_STEPS = REINDEX_STEP_SEQUENCE
  .reduce((steps, step) => ({
    ...steps,
    [step]: step,
  }), {});

export const UPGRADE_STEP_SEQUENCE = [
  'UPGRADE',
];

export const UPGRADE_STEPS = UPGRADE_STEP_SEQUENCE
  .reduce((steps, step) => ({
    ...steps,
    [step]: step,
  }), {});

export const NUMBER_OF_STEPS = {
  REINDEX: REINDEX_STEP_SEQUENCE.length,
  UPGRADE: UPGRADE_STEP_SEQUENCE.length,
};

export const STEP_RESULTS = {
  COMPLETED:   'COMPLETED',
  FAILED:      'FAILED',
  RUNNING:     'RUNNING',
  NOT_STARTED: 'NOT_STARTED',
  CANCELED:    'CANCELED',
};

export const INDEX_ACTION = {
  TYPE: {
    REINDEX: 'REINDEX',
    UPGRADE: 'UPGRADE',
  },
  LABEL: {
    REINDEX: 'Reindex',
    UPGRADE: 'Upgrade',
  },
};
