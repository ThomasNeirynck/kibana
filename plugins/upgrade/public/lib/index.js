export { timeout } from './util';
export {
  getAssistance,
  getMappingsAndSettings,
  createIndex,
  setReadOnly,
  runReindex,
  getTaskDetails,
  refreshIndex,
  verifyDocs,
  getSettingsAndAliases,
  updateRefreshInterval,
  replaceIndex,
  deleteTask,
  resetIndex,
  cancelTask,
  runUpgrade,
  isCompleted,
  isFailed,
  isNotStarted,
  isRunning,
  isStepCompleted,
  isStepFailed,
  isStepNotStarted,
  isStepRunning,
  isResettable,
  getStepMessage,
  wrapErrorMessage,
} from './reindex';
