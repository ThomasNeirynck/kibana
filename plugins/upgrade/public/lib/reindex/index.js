export {
  isCompleted,
  isFailed,
  isNotStarted,
  isRunning,
  isCanceled,
  isResettable,
  isCancelable,
  isStepCompleted,
  isStepFailed,
  isStepNotStarted,
  isStepRunning,
  isStepCanceled,
} from './state';

export {
  getStepMessage,
  wrapErrorMessage,
} from './messages';

export {
  getAssistance,
  getMappingsAndSettings,
  getSettingsAndAliases,
  updateRefreshInterval,
  createIndex,
  setReadOnly,
  runReindex,
  getTaskDetails,
  cancelTask,
  refreshIndex,
  verifyDocs,
  replaceIndex,
  deleteTask,
  resetIndex,
  runUpgrade,
} from './step_methods';

