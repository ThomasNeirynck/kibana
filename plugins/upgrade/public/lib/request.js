import chrome from 'ui/chrome';

function get(url) {
  return new Promise((resolve, reject) => {
    const req = new window.XMLHttpRequest();

    req.open('GET', url);

    req.onload = () => {
      if (req.status === 200) {
        resolve(JSON.parse(req.response));
      } else {
        reject(JSON.parse(req.response));
      }
    };

    req.onerror = () => {
      reject(new Error('Network error'));
    };

    req.send();
  });
}

export function getFromApi(url) {
  return get(chrome.addBasePath(url));
}

function post(url, data) {
  return new Promise((resolve, reject) => {
    const xhr = new window.XMLHttpRequest();
    xhr.open('POST', url);
    xhr.setRequestHeader('Content-type', 'application/json');
    xhr.setRequestHeader('kbn-version', chrome.getXsrfToken());
    xhr.onload = () => {
      if (xhr.status === 200) {
        resolve(JSON.parse(xhr.response));
      } else {
        reject(JSON.parse(xhr.response));
      }
    };
    xhr.send(JSON.stringify(data));
  });
}

export function postToApi(url, data) {
  return post(chrome.addBasePath(url), data);
}

function put(url, data) {
  return new Promise((resolve, reject) => {
    const xhr = new window.XMLHttpRequest();
    xhr.open('PUT', url);
    xhr.setRequestHeader('Content-type', 'application/json');
    xhr.setRequestHeader('kbn-version', chrome.getXsrfToken());
    xhr.onload = () => {
      if (xhr.status === 200) {
        resolve(JSON.parse(xhr.response));
      } else {
        reject(JSON.parse(xhr.response));
      }
    };
    xhr.send(JSON.stringify(data));
  });
}

export function putToApi(url, data) {
  return put(chrome.addBasePath(url), data);
}

function del(url) {
  return new Promise((resolve, reject) => {
    const xhr = new window.XMLHttpRequest();
    xhr.open('DELETE', url);
    xhr.setRequestHeader('Content-type', 'application/json');
    xhr.setRequestHeader('kbn-version', chrome.getXsrfToken());
    xhr.onload = () => {
      if (xhr.status === 200) {
        resolve(JSON.parse(xhr.response));
      } else {
        reject(JSON.parse(xhr.response));
      }
    };
    xhr.send();
  });
}

export function deleteFromApi(url) {
  return del(chrome.addBasePath(url));
}
