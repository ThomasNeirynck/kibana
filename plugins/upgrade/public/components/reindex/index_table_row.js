import React from 'react';

import { IndexActionSteps } from './index_action_steps';
import { INDEX_ACTION } from '../../lib/constants';
import {
  isResettable,
  isCancelable,
  isNotStarted,
} from '../../lib';


export function IndexTableRow({
  cancelAction,
  index,
  resetAction,
  processIndex,
}) {
  return (
    <tbody>
      <tr className="kuiTableRow">
        <td className="kuiTableRowCell">
          <div className="kuiTableRowCell__liner">
            { index.name }
          </div>
        </td>
        <td className="kuiTableRowCell">
          <div className="kuiTableRowCell__liner">
            <button
              className="kuiMenuButton kuiMenuButton--primary"
              disabled={ !isNotStarted(index) }
              onClick={ () => processIndex(index.name) }
            >
              { getActionButtonLabel(index.action) }
            </button>
          </div>
        </td>
        <td className="kuiTableRowCell kuiTableRowCell--alignRight">
          <div className="kuiTableRowCell__liner">
            <div className="kuiMenuButtonGroup kuiMenuButtonGroup--alignRight">
              <button
                className="kuiMenuButton kuiMenuButton--danger"
                disabled={ !isCancelable(index) }
                onClick={ () => cancelAction(index.name) }
              >
                Cancel
              </button>
              <button
                className="kuiMenuButton kuiMenuButton--basic"
                disabled={ !isResettable(index) }
                onClick={ () => resetAction(index.name) }
              >
                Reset
              </button>
            </div>
          </div>
        </td>
      </tr>
      {
        index.steps.length === 0
        ? null
        : (
          <tr className="kuiTableRow actionOutput">
            <td className="kuiTableRowCell kuiTableRowCell--wrap kuiTableRowCell--mergeTop" colSpan="3">
              <div className="kuiTableRowCell__liner">
                <IndexActionSteps
                  action={ index.action }
                  indexName={ index.name }
                  steps={ index.steps }
                />
              </div>
            </td>
          </tr>
        )
      }
    </tbody>
  );
}

IndexTableRow.propTypes = {
  cancelAction: React.PropTypes.func,
  index: React.PropTypes.object,
  resetAction: React.PropTypes.func,
  processIndex: React.PropTypes.func,
};

IndexTableRow.defaultProps = {
  cancelAction: () => {},
  index: {},
  resetAction: () => {},
  processIndex: () => {},
};

function getActionButtonLabel(action) {
  return INDEX_ACTION.LABEL[action];
  // TODO: might want to throw if assistance API
  // changes and doesn't give back actions we expect.
}
