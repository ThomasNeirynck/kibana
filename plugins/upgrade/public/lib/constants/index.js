import { HomeView } from '../../components/home';
import { CheckupView } from '../../components/checkup';
import { ReindexView } from '../../components/reindex';
import { LoggingView } from '../../components/logging';


export {
  NUMBER_OF_STEPS,
  REINDEX_STEPS,
  UPGRADE_STEPS,
  STEP_RESULTS,
  INDEX_ACTION,
} from './reindex';

export {
  DEPRECATION_ISSUE_LEVELS,
} from './deprecations';

export const VIEWS = {
  HOME: {
    component: HomeView,
    label: 'Overview',
    location: 'home',
  },
  CHECKUP: {
    component: CheckupView,
    label: 'Cluster Checkup',
    location: 'checkup',
  },
  REINDEX: {
    component: ReindexView,
    label: 'Reindex Helper',
    location: 'reindex',
  },
  LOGGING: {
    component: LoggingView,
    label: 'Toggle Deprecation Logging',
    location: 'logging',
  }
};
export const DEFAULT_VIEW_ID = 'HOME';

export const HUMAN_READABLE_DELAY = 750;

export const LOADING_STATUS = {
  FORBIDDEN: 'FORBIDDEN',
  FAILURE: 'FAILURE',
  LOADING: 'LOADING',
  SUCCESS: 'SUCCESS',
  UNINITIALIZED: 'UNINITIALIZED',
};
