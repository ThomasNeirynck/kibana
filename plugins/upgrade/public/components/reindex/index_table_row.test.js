import { shallow } from 'enzyme';
import React from 'react';

import { IndexTableRow } from './index_table_row';
import { isCancelable, isNotStarted, isResettable } from '../../lib';
import { INDEX_ACTION } from '../../lib/constants';

jest.mock('ui/chrome', () => {}, { virtual: true });
jest.mock('ui/notify/notifier', () => ({
  Notifier: function Notifier() {},
}), { virtual: true });
jest.mock('../../lib/reindex/state');


describe('IndexTableRow', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  test('renders the index name', () => {
    const component = (
      <IndexTableRow
        index={{
          action: INDEX_ACTION.TYPE.UPGRADE,
          name: 'A UNIQUE INDEX NAME',
          steps: [],
        }}
      />
    );

    const wrapper = shallow(component);

    expect(wrapper.containsMatchingElement('A UNIQUE INDEX NAME')).toBe(true);
  });

  describe('upgrade/reindex button', () => {
    test('is enabled for indices that require an upgrade', () => {
      isNotStarted.mockReturnValue(true);
      const component = (
        <IndexTableRow
          index={{
            action: INDEX_ACTION.TYPE.UPGRADE,
            name: 'INDEX NAME',
            steps: [],
          }}
        />
      );

      expect(shallow(component).containsMatchingElement(
        <button disabled={false} >{ INDEX_ACTION.LABEL.UPGRADE }</button>
      )).toBe(true);
    });

    test('is disabled for indices that are being upgraded', () => {
      isNotStarted.mockReturnValue(false);
      const component = (
        <IndexTableRow
          index={{
            action: INDEX_ACTION.TYPE.UPGRADE,
            name: 'INDEX NAME',
            steps: [],
          }}
        />
      );

      expect(shallow(component).containsMatchingElement(
        <button disabled={true}>{ INDEX_ACTION.LABEL.UPGRADE }</button>
      )).toBe(true);
    });

    test('is enabled for indices that require reindexing', () => {
      isNotStarted.mockReturnValue(true);
      const component = (
        <IndexTableRow
          index={{
            action: INDEX_ACTION.TYPE.REINDEX,
            name: 'INDEX NAME',
            steps: [],
          }}
        />
      );

      expect(shallow(component).containsMatchingElement(
        <button disabled={false} >{ INDEX_ACTION.LABEL.REINDEX }</button>
      )).toBe(true);
    });

    test('is disabled for indices that are being reindexed', () => {
      isNotStarted.mockReturnValue(false);
      const component = (
        <IndexTableRow
          index={{
            action: INDEX_ACTION.TYPE.REINDEX,
            name: 'INDEX NAME',
            steps: [],
          }}
        />
      );

      expect(shallow(component).containsMatchingElement(
        <button disabled={true}>{ INDEX_ACTION.LABEL.REINDEX }</button>
      )).toBe(true);
    });

    test('calls the processIndex property when clicked', () => {
      isNotStarted.mockReturnValue(true);
      const processIndex = jest.fn();
      const component = (
        <IndexTableRow
          index={{
            action: INDEX_ACTION.TYPE.UPGRADE,
            name: 'INDEX',
            steps: [],
          }}
          processIndex={processIndex}
        />
      );

      const wrapper = shallow(component);
      wrapper.findWhere((node) => node.matchesElement(
        <button>{ INDEX_ACTION.LABEL.UPGRADE }</button>
      )).simulate('click');

      expect(processIndex).toHaveBeenCalledWith('INDEX');
    });
  });

  describe('cancel button', () => {
    test('is enabled for indices that are being processed', () => {
      isCancelable.mockReturnValue(true);
      const component = (
        <IndexTableRow
          index={{
            action: INDEX_ACTION.TYPE.UPGRADE,
            name: 'INDEX NAME',
            steps: [],
          }}
        />
      );

      expect(shallow(component).containsMatchingElement(
        <button disabled={false} >Cancel</button>
      )).toBe(true);
    });

    test('is disabled for indices that are not being processed', () => {
      isCancelable.mockReturnValue(false);
      const component = (
        <IndexTableRow
          index={{
            action: INDEX_ACTION.TYPE.UPGRADE,
            name: 'INDEX NAME',
            steps: [],
          }}
        />
      );

      expect(shallow(component).containsMatchingElement(
        <button disabled={true} >Cancel</button>
      )).toBe(true);
    });

    test('calls the cancelAction property when clicked', () => {
      isCancelable.mockReturnValue(true);
      const cancelAction = jest.fn();
      const component = (
        <IndexTableRow
          index={{
            action: INDEX_ACTION.TYPE.UPGRADE,
            name: 'INDEX',
            steps: [],
          }}
          cancelAction={cancelAction}
        />
      );

      const wrapper = shallow(component);
      wrapper.findWhere((node) => node.matchesElement(
        <button>Cancel</button>
      )).simulate('click');

      expect(cancelAction).toHaveBeenCalledWith('INDEX');
    });
  });

  describe('reset button', () => {
    test('is enabled for indices that are resettable', () => {
      isResettable.mockReturnValue(true);
      const component = (
        <IndexTableRow
          index={{
            action: INDEX_ACTION.TYPE.UPGRADE,
            name: 'INDEX NAME',
            steps: [],
          }}
        />
      );

      expect(shallow(component).containsMatchingElement(
        <button disabled={false} >Reset</button>
      )).toBe(true);
    });

    test('is disabled for indices that are not resettable', () => {
      isResettable.mockReturnValue(false);
      const component = (
        <IndexTableRow
          index={{
            action: INDEX_ACTION.TYPE.UPGRADE,
            name: 'INDEX NAME',
            steps: [],
          }}
        />
      );

      expect(shallow(component).containsMatchingElement(
        <button disabled={true} >Reset</button>
      )).toBe(true);
    });

    test('calls the resetAction property when clicked', () => {
      isResettable.mockReturnValue(true);
      const resetAction = jest.fn();
      const component = (
        <IndexTableRow
          index={{
            action: INDEX_ACTION.TYPE.UPGRADE,
            name: 'INDEX',
            steps: [],
          }}
          resetAction={resetAction}
        />
      );

      const wrapper = shallow(component);
      wrapper.findWhere((node) => node.matchesElement(
        <button>Reset</button>
      )).simulate('click');

      expect(resetAction).toHaveBeenCalledWith('INDEX');
    });
  });

  test('renders no IndexActionSteps component when there are no steps', () => {
    const component = (
      <IndexTableRow
        index={{
          action: INDEX_ACTION.TYPE.UPGRADE,
          name: 'INDEX',
          steps: [],
        }}
      />
    );

    const wrapper = shallow(component);

    expect(wrapper.find('IndexActionSteps').exists()).toBe(false);
  });

  test('renders an IndexActionSteps component when there are steps', () => {
    const component = (
      <IndexTableRow
        index={{
          action: INDEX_ACTION.TYPE.UPGRADE,
          name: 'INDEX',
          steps: [
            { name: 'STEP 1' },
          ],
        }}
      />
    );

    const wrapper = shallow(component);

    expect(wrapper.find('IndexActionSteps')).toMatchSnapshot();
  });

});
