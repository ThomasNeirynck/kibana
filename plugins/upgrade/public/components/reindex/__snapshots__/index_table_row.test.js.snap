// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IndexTableRow renders an IndexActionSteps component when there are steps 1`] = `
<IndexActionSteps
  action="UPGRADE"
  indexName="INDEX"
  steps={
    Array [
      Object {
        "name": "STEP 1",
      },
    ]
  }
/>
`;
